src/main/java/net/openhft/compiler/CachedCompiler.java

```diff
@@ -77,67 +90,58 @@ public void close() {
-        }
-        // reuse the same file manager to allow caching of jar files
-        boolean ok = s_compiler.getTask(writer, fileManager, new DiagnosticListener<JavaFileObject>() {
-            @Override
-            public void report(Diagnostic<? extends JavaFileObject> diagnostic) {
-                if (diagnostic.getKind() == Diagnostic.Kind.ERROR) {
-                    writer.println(diagnostic);
-                }
-            }
-        }, options, null, compilationUnits).call();
-
-        if (!ok) {
-            // compilation error, so we want to exclude this file from future compilation passes
-            if (sourceDir == null)
-                javaFileObjects.remove(className);
-
-            // nothing to return due to compiler error
-            return Collections.emptyMap();
-        } else {
-            Map<String, byte[]> result = fileManager.getAllBuffers();
-
-            return result;
-        }
+    /**
+     * Delegates to {@link #loadFromJava(ClassLoader, String, String, PrintWriter, DiagnosticListener)}.
+     * Only error diagnostics are collected, and are written to {@code writer}.
+     */
+    public Class<?> loadFromJava(@NotNull ClassLoader classLoader,
+            @NotNull String className,
+            @NotNull String javaCode,
+            @Nullable PrintWriter writer) throws ClassNotFoundException {
+        return loadFromJava(classLoader, className, javaCode, writer, null);
     }
 
+    /**
+     * Gets a previously compiled and loaded class, or compiles the given Java code and
+     * loads the class.
+     *
+     * @param classLoader class loader for defining and loading the class
+     * @param className binary name of the class to load, for example {@code com.example.MyClass$Nested}
+     * @param javaCode Java code to compile, in case the class had not been compiled and loaded before
+     * @param writer writer for compilation information and diagnostics (should be thread-safe);
+     *      when {@code null} defaults to writing to {@link System#err}
+     * @param diagnosticListener listener for diagnostics emitted by the compiler (should be thread-safe);
+     *      when {@code null}, error diagnostics are written to the {@code writer}, other diagnostics are ignored
+     * @return the loaded class
+     * @throws ClassNotFoundException if compiling or loading the class failed; inspect {@code writer} or
+     *      {@code diagnosticListener} for additional details
+     */
     public Class<?> loadFromJava(@NotNull ClassLoader classLoader,
                               @NotNull String className,
                               @NotNull String javaCode,
-                              @Nullable PrintWriter writer) throws ClassNotFoundException {
+                              @Nullable PrintWriter writer,
+                              @Nullable DiagnosticListener<? super JavaFileObject> diagnosticListener) throws ClassNotFoundException {
```

Comment:
Not completely sure about the `? super JavaFileObject`; maybe users make assumptions then about the `JavaFileObject` of the diagnostics, which are actually implementation details of Java-Runtime-Compiler, e.g. the file paths.